crates/test-programs/src/bin/async_cancel_callee.rs

```diff
@@ -0,0 +1,231 @@
+    unreachable!()
+}
+
+const ON_CANCEL_TASK_RETURN: u32 = 0;
+const ON_CANCEL_TASK_CANCEL: u32 = 1;
+
+#[derive(Clone, Copy)]
+struct SleepParams {
+    time_in_millis: u64,
+    on_cancel: u32,
+    on_cancel_delay_millis: u64,
+    synchronous_delay: bool,
+}
+
+enum State {
+    S0(SleepParams),
+    S1 {
+        set: u32,
+        waitable: u32,
+        params: SleepParams,
+        param_ptr: *mut u64,
+    },
+    S2 {
+        set: u32,
+        waitable: u32,
+        params: SleepParams,
+        param_ptr: *mut u64,
+    },
+}
+
+#[unsafe(export_name = "local:local/backpressure#set-backpressure")]
+unsafe extern "C" fn export_set_backpressure(enabled: bool) {
+    async_support::backpressure_set(enabled);
+}
+
+#[unsafe(export_name = "local:local/sleep#sleep-millis")]
+unsafe extern "C" fn export_sleep_sleep_millis(time_in_millis: u64) {
+    unsafe {
+        sleep_millis(time_in_millis);
+    }
+}
+
+#[unsafe(export_name = "[async-lift]local:local/sleep-with-options#sleep-millis")]
+unsafe extern "C" fn export_sleep_with_options_sleep_millis(
+    time_in_millis: u64,
+    on_cancel: u32,
+    on_cancel_delay_millis: u64,
+    synchronous_delay: bool,
+) -> u32 {
+    unsafe {
+        context_set(
+            u32::try_from(Box::into_raw(Box::new(State::S0(SleepParams {
+                time_in_millis,
+                on_cancel,
+                on_cancel_delay_millis,
+                synchronous_delay,
+            }))) as usize)
+            .unwrap(),
+        );
+        callback_sleep_with_options_sleep_millis(EVENT_NONE, 0, 0)
+    }
+}
+
+#[unsafe(export_name = "[callback][async-lift]local:local/sleep-with-options#sleep-millis")]
+unsafe extern "C" fn callback_sleep_with_options_sleep_millis(
+    event0: u32,
+    event1: u32,
+    event2: u32,
+) -> u32 {
+    unsafe {
+        let state = &mut *(usize::try_from(context_get()).unwrap() as *mut State);
+        match state {
+            State::S0(params) => {
+                assert_eq!(event0, EVENT_NONE);
+
+                let param_ptr = Box::into_raw(Box::new(params.time_in_millis));
+
+                let status = sleep_millis_async(param_ptr.cast(), ptr::null_mut());
+
+                let waitable = status >> 4;
+                let status = status & 0xF;
+
+                assert_eq!(status, STATUS_STARTED);
+
+                let set = waitable_set_new();
+                waitable_join(waitable, set);
+
+                *state = State::S1 {
+                    set,
+                    waitable,
+                    params: *params,
+                    param_ptr,
+                };
+
+                CALLBACK_CODE_WAIT | (set << 4)
+            }
+
+            State::S1 {
+                set,
+                waitable,
+                params,
+                param_ptr,
+            } => {
+                assert_eq!(event0, EVENT_CANCELLED);
+
+                let result = subtask_cancel(*waitable);
+
+                assert_eq!(result, DONE);
```

Comment:
Instead of "DONE" vs "BLOCKED", I think this'll need to return the state of the task itself. For example in guest bindings it needs to differentiate between `STATUS_{START,RETURN}_CANCELLED` to determine what needs to be deallocated. Would it be possible to change the ABI here so `subtask.cancel` returns the current status of the task?

Reply:
Yeah, I chatted with Luke about that earlier this afternoon.  Away from keyboard now, but will update it when I get a chance.